lib/host/wasi_nn/wasinnfunc.cpp

```diff
@@ -3,49 +3,478 @@
+                                  uint32_t Encoding,
                                   uint32_t Target [[maybe_unused]],
                                   uint32_t GraphPtr [[maybe_unused]]) {
-  return 0;
+  // GraphBuilders' Layout: |builder-0|builder-0 len|builder-1|builder-1 len|...
+  // Check memory instance from module.
+  if (MemInst == nullptr) {
+    return Unexpect(ErrCode::ExecutionFailed);
+  }
+
+  if (Encoding == Ctx.BackendsMapping.at("OpenVINO")) {
+#ifdef WASMEDGE_WASINN_BUILD_OPENVINO
+    uint32_t *GraphBuilders = MemInst->getPointer<uint32_t *>(BuilderPtr, 1);
+    uint32_t *GraphId = MemInst->getPointer<uint32_t *>(GraphPtr, 1);
+
+    if (BuilderLen != 2) {
+      spdlog::error("[WASI-NN] Wrong GraphBuilder Length {:d}, expecting 2",
+                    BuilderLen);
+      return static_cast<uint32_t>(WASINN::ErrNo::InvalidArgument);
+    }
+    IEStatusCode Status;
+    if (Ctx.OpenVINOCore == nullptr) {
+      Status = ie_core_create("", &Ctx.OpenVINOCore);
+      if (Status != IEStatusCode::OK || Ctx.OpenVINOCore == nullptr) {
+        spdlog::error("[WASI-NN] Error happened when init OpenVINO core.");
+        return static_cast<uint32_t>(WASINN::ErrNo::MissingMemory);
+      }
+      spdlog::debug("[WASI-NN] Initialize OpenVINO Core");
+    }
+
+    std::string DeviceName = mapTargetToString(Target);
+    if (DeviceName.length() == 0) {
+      spdlog::error("[WASI-NN] Device target {:d} not support!", Target);
+      return static_cast<uint32_t>(WASINN::ErrNo::InvalidArgument);
+    } else {
+      spdlog::debug("[WASI-NN] Using device: {:s}", DeviceName);
+    }
+
+    uint32_t XMLStringLen = GraphBuilders[1];
+    uint32_t WeightBinsLen = GraphBuilders[3];
+    uint8_t *XMLPtr = MemInst->getPointer<uint8_t *>(GraphBuilders[0], 1);
+    uint8_t *BinPtr = MemInst->getPointer<uint8_t *>(GraphBuilders[2], 1);
+    std::vector<uint8_t> XMLStrings(XMLPtr, XMLPtr + XMLStringLen);
+    std::vector<uint8_t> WeightBins(BinPtr, BinPtr + WeightBinsLen);
+    spdlog::debug("[WASI-NN] read xml length {:d}", XMLStrings.size());
+    spdlog::debug("[WASI-NN] read bin length {:d}", WeightBins.size());
+
+    tensor_desc_t WeightsDesc{
+        layout_e::ANY, {1, {WeightBins.size()}}, precision_e::U8};
+    ie_blob_t *WeightsBlob = nullptr;
+    ie_blob_buffer_t BlobCBuffer;
+
+    Status = ie_blob_make_memory(&WeightsDesc, &WeightsBlob);
+    if (Status != IEStatusCode::OK || WeightsBlob == nullptr) {
+      spdlog::error(
+          "[WASI-NN] Unable to create model's weight blob, error code: {}",
+          Status);
+      return static_cast<uint32_t>(WASINN::ErrNo::Busy);
+    }
+    Status = ie_blob_get_cbuffer(WeightsBlob, &BlobCBuffer);
+    if (Status != IEStatusCode::OK) {
+      spdlog::error(
+          "[WASI-NN] Unable to find weight blob's buffer, error code: {}",
+          Status);
+      return static_cast<uint32_t>(WASINN::ErrNo::MissingMemory);
+    }
+    uint8_t *BlobData = const_cast<uint8_t *>(
```

Comment:
The `ie_blob_buffer` is a `union`:
https://docs.openvino.ai/cn/2021.1/ie_c_api/structie__blob__buffer.html

Why don't you use the `BlobCBuffer.buffer` directly instead of using `const_cast`?

Reply:
Got it, I didn't survey the `ie_blob_buffer` clearly...